### PR TITLE
Draft of adding stochasticity

### PR DESCRIPTION
**Pull request by FiodorG**
Left to do:
- [x] Confirm with modelers about the model choices
- [ ] Tests
- [x] Check that stochastic model has only int values
- [x] Check runs with multiple trials are all different (seeding is done well)
- [x] Confirm deterministic run is as fast
- [x] Pass in random_state to all deterministic calls for reproducibility

### CODE REVIEW

File: simple_network_sim/network_of_populations.py [hunk 1]
```diff
@@ -412,20 +415,30 @@ def getInternalInfectiousContactsInNode(
...
+            for ageFrom in mixingMatrix[ageTo]:
+
+                if RANDOM:
+                    infectious = getInfectious(ageFrom, currentInternalStateDict, infectiousStates)
+
+                    contactsAverage = mixingMatrix[ageFrom][ageTo] * contactsMultiplier
+                    numberOfContacts = stats.poisson.rvs(contactsAverage, size=int(infectious))
+                    numberOfContacts = numberOfContacts[numberOfContacts > 0]
+
+                    infectiousContacts[ageTo] += np.sum(stats.hypergeom.rvs(int(totalInAge), int(susceptibles), numberOfContacts))
```
Comment (aflag) [inline simple_network_sim/network_of_populations.py:434]: Why are you using hypergeom here?

Comment (FiodorG) [reply simple_network_sim/network_of_populations.py:434]: So this is from my understanding of what the mixing matrix means.
1) If it's "number of contacts with distinct different people", and the value is 3, we need to pick 3 individuals WITHOUT replacement in the population, and see if we select susceptible. 
2) If it's "number of contacts with people, but they could be the same", and the value is 3, we need to pick 3 individuals WITH replacement in the population, and see if we select susceptible. 

Hypergeometric is basically like the binomial distribution but without replacement. I assumed 1) here, but was going to ask about this at the moment of the proper review.

File: simple_network_sim/network_of_populations.py [hunk 2]
```diff
@@ -31,40 +33,7 @@ class NetworkOfPopulation(NamedTuple):
     movementMultipliers: Dict[Time, loaders.Multiplier]
     infectiousStates: List[Compartment]
     infectionProb: Dict[Time, float]
-
-
-# NotCurrentlyInUse
```
Comment (FiodorG) [inline simple_network_sim/network_of_populations.py:36]: Looks like we won't use it anymore? I can add it back if necessary

File: simple_network_sim/network_of_populations.py (hunk 1 above)
Comment (aflag) [reply simple_network_sim/network_of_populations.py]: So you're saying that it counts the number of distinct contacts rather than total number of contacts (which could include meeting one person more than once), right? I think you're right, I think the question they asked was how many people you've contacted.

Comment (aflag) [reply simple_network_sim/network_of_populations.py]: > We will now ask you to remember who you have been in contact with yesterday, between 5am yesterday and 5am today. We are only interested in direct contacts, which are people who you met in personand with whomyou exchanged at least a few words, or with whomyou had physical contact (e.g.a handshake, embracing, kissing, contact sports).Note that if you only spoke to someone over the phone or internet, they should not be included in this section. 

https://www.medrxiv.org/content/medrxiv/suppl/2020/04/03/2020.03.31.20049023.DC1/2020.03.31.20049023-2.pdf

So the people interviewed in the comix study would tell them a list of people they met the day before.

File: simple_network_sim/network_of_populations.py (hunk 2 above)
Comment (aflag) [reply simple_network_sim/network_of_populations.py:36]: I suggest we don't do changes like this in the same PR as other changes. This will cause a conflict with another change where I turned this into a package, for instance.

File: simple_network_sim/network_of_populations.py (hunk 1 above)
Comment (FiodorG) [reply simple_network_sim/network_of_populations.py]: Great, thanks a lot for confirming this.

File: simple_network_sim/network_of_populations.py (hunk 2 above)
Comment (FiodorG) [reply simple_network_sim/network_of_populations.py:36]: ok, reverted the change

File: simple_network_sim/network_of_populations.py (hunk 1 above)
Comment (magicicada) [reply simple_network_sim/network_of_populations.py]: Thanks for checking on this - I was just checking it, and agree.  Hypergeometric is the right choice here for sampling without replacement.

**Comment by magicicada** [comment]
@aflag Is the swap in the mixing matrix storage going to cause any problems?  I assume not, but thought I'd confirm.

**Comment by aflag** [comment]
I only did that because it improved performance for the deterministic code, but it probably would make the stochastic slower. Other than that, there shouldn't be any behavioral changes.

File: simple_network_sim/network_of_populations.py [hunk 3]
```diff
@@ -231,38 +236,72 @@ def getTotalSuscept(nodeState):
...
+    :param newContacts: The number of new contacts to be distributed across age ranges.
+    :type newContacts: int
+    :param stochastic: Whether to run the model in a stochastic or deterministic mode
+    :type stochastic: bool
+    :return: The number of new infections in each age group.
+    :rtype: A dictionary of ages (keys) and the number of new infections (values)
+    """
+    if stochastic:
+        ageProbabilities = np.array(list(ageToSusceptibles.values())) / totalSusceptibles
+        allocationsByAge = stats.multinomial.rvs(newContacts, ageProbabilities)
```
Comment (aflag) [inline simple_network_sim/network_of_populations.py:297]: I don't have much experience with the scipy.stats module. But all functions seem to take a random number generator as a paramter. Could we store the generator in the NetworkOfPopulations tuple and use the same generator everywhere? I think the seed used needs to be a paramter or an output of the model as well (so that people can reproduce runs). I think that will affect the way we parelelise code too, but, even if we run it across different machines, we can still prepare a set of seeds ahead of time.

Comment (FiodorG) [reply simple_network_sim/network_of_populations.py:297]: I agree with you that we need to store the seed for reproducibility. However, there are many pitfalls here. 
1) We need a different seed for every trial (if we use the same seed for every trial every trial will be identical). Using random seeds is not good either, as they might still be repeated out of pure luck. Usually using sequential seeds is the way to go (see https://www.johndcook.com/blog/2016/01/29/random-number-generator-seed-mistakes/)
2) Randomness is not guaranteed for numbers generated by different seeds! Only for number generated sequentially out of a single seed (see e.g. https://blogs.unity3d.com/2015/01/07/a-primer-on-repeatable-random-numbers/). So ideally we set the seed once and for all before doing all the trials SEQUENTIALLY.

For parallelizing the random trials, I will need to think a bit about it.

Comment (aflag) [reply simple_network_sim/network_of_populations.py:297]: Regarding passing the same generator to all methods, that's why I suggested putitng it in the NetworkOfPopulations tuple. My expectation is that we are going to parallelise the multiple runs of the model, not internally (I've did a few attempls of internal parallelisation and that didn't really to help, it's tricky to avoid copying a bunch of data around). So, we should always consider the whole model is always going to be single threaded.

Now, generating multiple seeds for each run, I don't know what's the best way to do it. Don't we just need to guarantee that each run has a different seed value?

Comment (FiodorG) [reply simple_network_sim/network_of_populations.py]: Thanks, sounds good for the single threaded model. I've just made a commit to initialize a random_state generator at network creation.

For multiple seeds on each run, theoretically it's not quite correct (as per my second link value, they show an example of a series of n numbers from one seed vs 1 number from n seeds). But in practice lot of places do like this. I'll try to look around to see how it's usually done.

**Comment by FiodorG** [comment]
Example run for one HB: 
image

File: simple_network_sim/network_of_populations.py (hunk 3 above)
Comment (aflag) [reply simple_network_sim/network_of_populations.py]: Maybe I'm misunderstanding something, but the blog post you pasted in the first link said exaclty what I said:

> In this case it would have been better to seed each process with sequential seeds: give the first process seed 1, the second seed 2, etc. The seeds don’t have to be random; they just have to be unique. If you’re using a good random number generator, the outputs of 1,000 processes seeded with 1, 2, 3, …, 1000 will be independent.

That's our scenario. The only reason I wanted to use uuids or something like that is because we want the ability of running different simulations, rather than always use the same fixed seeds. Although always keeping the same seeds may be ok too (as we don't expect much difference between runs of thousands of concurrent models).

**Comment by aflag** [comment]
@magicicada btw, I should've mentioned in my comment above. He's actually reverting a change that I made, the current order is the original one

File: simple_network_sim/network_of_populations.py (hunk 3 above)
Comment (FiodorG) [reply simple_network_sim/network_of_populations.py]: Yes sorry for not more clear, that's what most practitioners do (and most places I know of/been to). But in theory this isn't correct (the second blog post shows that), as random numbers generated from different seeds are not independent from each other. So I think we can start by giving sequential seeds is fine for now, but I will check how that problem is usually solved otherwise.

File: simple_network_sim/loaders.py [hunk 4]
```diff
@@ -114,7 +108,7 @@ def genGraphFromContactFile(commutes: pd.DataFrame) -> nx.DiGraph:
     Pairs of nodes are listed in the file by source, destination, weight and adjustment.
 
     :param commutes: Weighted edge list.
-    :type pd.DataFrame: pandas DataFrame.
+    :type commutes: pandas DataFrame.
```
Comment (aflag) [inline simple_network_sim/loaders.py:111]: I think the type was actually correct before

File: simple_network_sim/network_of_populations.py (hunk 3 above)
Comment (aflag) [reply simple_network_sim/network_of_populations.py]: I think I understand your point now. I was considering different runs as independent runs. In that case, we wouldn't care that two random sequences are not independent from each other. But they are not actually independent because we are going to average out in the end, right? I think to get around that problem we'd have to either write or find a very sophisticated random number generator which is able to create multiple independent sequences.

File: simple_network_sim/sampleUseOfModel.py [hunk 5]
```diff
@@ -77,9 +77,8 @@ def runSimulation(
     if trials <= 1:
         # The averaging logic is slow and wastes a bunch of memory, skip it if we don't need it
         logger.info("Running simulation (1/1)")
-        disposableNetwork = copy.deepcopy(network)
```
Comment (aflag) [inline simple_network_sim/sampleUseOfModel.py:80]: I suggest you don't delete this just yet. Maybe we can make the CLI changes in another PR.

File: simple_network_sim/network_of_populations.py (hunk 3 above)
Comment (FiodorG) [reply simple_network_sim/network_of_populations.py]: Yes exactly, we average things that are not quite independent so the expectation will be slightly off. But given it seems it's a problem that half the world would have, even if we don't do anything special about it in the end, we should keep in mind and have a bit of a look around at least. 

I've known about that fact for some time but never knew if there was a solution to that.

File: simple_network_sim/sampleUseOfModel.py (hunk 5 above)
Comment (FiodorG) [reply simple_network_sim/sampleUseOfModel.py]: Done
